pilot/pkg/simulation/traffic.go

```diff
@@ -0,0 +1,434 @@
+	sim := &Simulation{
+		t:         t,
+		Listeners: s.Listeners(proxy),
+		Clusters:  s.Clusters(proxy),
+		Routes:    s.Routes(proxy),
+	}
+	return sim
+}
+
+func NewSimulation(t *testing.T, s *xds.FakeDiscoveryServer, proxy *model.Proxy) *Simulation {
+	return NewSimulationFromConfigGen(t, s.ConfigGenTest, proxy)
+}
+
+// withT swaps out the testing struct. This allows executing sub tests.
+func (sim *Simulation) withT(t *testing.T) *Simulation {
+	cpy := *sim
+	cpy.t = t
+	return &cpy
+}
+
+func (sim *Simulation) RunExpectations(es []Expect) {
+	for _, e := range es {
+		sim.t.Run(e.Name, func(t *testing.T) {
+			sim.withT(t).Run(e.Call).Matches(t, e.Result)
+		})
+	}
+}
+
+func (sim *Simulation) Run(input Call) (result Result) {
+	result = Result{t: sim.t}
+	input = input.FillDefaults()
+
+	// First we will match a listener
+	l := matchListener(sim.Listeners, input)
+	if l == nil {
+		result.Error = ErrNoListener
+		return
+	}
+	result.ListenerMatched = l.Name
+
+	// Apply listener filters. This will likely need the TLS inspector in the future as well
+	if _, f := xdstest.ExtractListenerFilters(l)[xdsfilters.HTTPInspector.Name]; f {
+		input.Alpn = protocolToAlpn(input.Protocol)
+	}
+	fc, err := sim.matchFilterChain(l.FilterChains, input)
+	if err != nil {
+		result.Error = err
+		return
+	}
+	result.FilterChainMatched = fc.Name
+
+	if hcm := xdstest.ExtractHTTPConnectionManager(sim.t, fc); hcm != nil {
+		routeName := hcm.GetRds().RouteConfigName
+		result.RouteConfigMatched = routeName
+		rc := xdstest.ExtractRouteConfigurations(sim.Routes)[routeName]
+		if len(input.Headers["Host"]) != 1 {
+			result.Error = errors.New("http requests require a host header")
+			return
+		}
+		vh := sim.matchDomain(rc, input.Headers["Host"][0])
+		if vh == nil {
+			result.Error = ErrNoVirtualHost
+			return
+		}
+		result.VirtualHostMatched = vh.Name
+		r := sim.matchVirtualHost(vh, input)
+
+		if r == nil {
+			result.Error = ErrNoRoute
+			return
+		}
+		result.RouteMatched = r.Name
+		switch t := r.GetAction().(type) {
+		case *route.Route_Route:
+			result.ClusterMatched = t.Route.GetCluster()
+		}
+	} else if tcp := xdstest.ExtractTCPProxy(sim.t, fc); tcp != nil {
+		result.ClusterMatched = tcp.GetCluster()
+	}
+	return
+}
+
+func (sim *Simulation) matchVirtualHost(vh *route.VirtualHost, input Call) *route.Route {
+	for _, r := range vh.Routes {
+		// check path
+		switch pt := r.Match.GetPathSpecifier().(type) {
+		case *route.RouteMatch_Prefix:
+			if !strings.HasPrefix(input.Path, pt.Prefix) {
+				continue
+			}
+		case *route.RouteMatch_Path:
+			if input.Path != pt.Path {
+				continue
+			}
+		case *route.RouteMatch_SafeRegex:
+			r, err := regexp.Compile(pt.SafeRegex.GetRegex())
+			if err != nil {
+				sim.t.Fatalf("invalid regex %v: %v", r, err)
+			}
+			if !r.MatchString(input.Path) {
+				continue
+			}
+		default:
+			sim.t.Fatalf("unknown route path type")
+		}
+
```

Comment:
nit you are missing header matches :). You dont have to add it here but just add a TODO.

Reply:
added a todo